plugins/outputs/postgresql/table_manager.go

```diff
@@ -0,0 +1,388 @@
+			if err := tm.executeTemplates(ctx, db, createTemplates, tbl, columns, metricsTable, tagsTable); err != nil {
+				tbl.Unlock()
+				return nil, fmt.Errorf("creating table: %w", err)
+			}
+
+			tblColumns = tbl.Columns()
+		}
+	}
+	tbl.Unlock()
+
+	missingColumns, err := tm.checkColumns(tblColumns, columns)
+	if err != nil {
+		return nil, fmt.Errorf("column validation: %w", err)
+	}
+	if len(missingColumns) == 0 {
+		return nil, nil
+	}
+
+	if len(addColumnsTemplates) == 0 {
+		return missingColumns, nil
+	}
+
+	tbl.Lock()
+	// Check again in case someone else got it while table was unlocked.
+	tblColumns = tbl.Columns()
+	missingColumns, _ = tm.checkColumns(tblColumns, columns)
+	if len(missingColumns) == 0 {
+		tbl.Unlock()
+		return nil, nil
+	}
+
+	if err := tm.executeTemplates(ctx, db, addColumnsTemplates, tbl, missingColumns, metricsTable, tagsTable); err != nil {
+		tbl.Unlock()
+		return nil, fmt.Errorf("adding columns: %w", err)
+	}
+	tbl.Unlock()
+	return tm.checkColumns(tbl.Columns(), columns)
+}
+
+func (tm *TableManager) checkColumns(dbColumns map[string]utils.Column, srcColumns []utils.Column) ([]utils.Column, error) {
+	var missingColumns []utils.Column
+	for _, srcCol := range srcColumns {
+		dbCol, ok := dbColumns[srcCol.Name]
+		if !ok {
+			missingColumns = append(missingColumns, srcCol)
+			continue
+		}
+		if !utils.PgTypeCanContain(dbCol.Type, srcCol.Type) {
+			return nil, fmt.Errorf("column type '%s' cannot store '%s'", dbCol.Type, srcCol.Type)
+		}
+	}
+	return missingColumns, nil
+}
+
+func (tm *TableManager) executeTemplates(
+	ctx context.Context,
+	db dbh,
+	tmpls []*template.Template,
+	tbl *tableState,
+	newColumns []utils.Column,
+	metricsTable *tableState,
+	tagsTable *tableState,
+) error {
+	tmplTable := template.NewTable(tm.Schema, tbl.name, colMapToSlice(tbl.Columns()))
+	metricsTmplTable := template.NewTable(tm.Schema, metricsTable.name, colMapToSlice(metricsTable.Columns()))
+	var tagsTmplTable *template.Table
+	if tagsTable != nil {
+		tagsTmplTable = template.NewTable(tm.Schema, tagsTable.name, colMapToSlice(tagsTable.Columns()))
+	} else {
+		tagsTmplTable = template.NewTable("", "", nil)
+	}
+
+	/* https://github.com/jackc/pgx/issues/872
+	stmts := make([]string, len(tmpls))
+	batch := &pgx.Batch{}
+	for i, tmpl := range tmpls {
+		sql, err := tmpl.Render(tmplTable, newColumns, metricsTmplTable, tagsTmplTable)
+		if err != nil {
+			return err
+		}
+		stmts[i] = string(sql)
+		batch.Queue(stmts[i])
+	}
+
+	batch.Queue(refreshTableStructureStatement, tm.Schema, tableName)
+
+	batchResult := tm.db.SendBatch(ctx, batch)
+	defer batchResult.Close()
+
+	for i := 0; i < len(tmpls); i++ {
+		if x, err := batchResult.Exec(); err != nil {
+			return fmt.Errorf("executing `%.40s...`: %v %w", stmts[i], x, err)
+		}
+	}
+
+	rows, err := batchResult.Query()
+	if err != nil {
+		return fmt.Errorf("refreshing table: %w", err)
+	}
+	tm.refreshTableStructureResponse(tableName, rows)
+	*/
```

Comment:
Is this needed?

Reply:
no. preferred, but due to limitations in pgx, it doesn't work. will remove